protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,489 @@
+            rdata.rewardRate = uint216((amount + leftover) / REWARD_DURATION);
+        }
+
+        rdata.lastUpdateTime = uint40(block.timestamp);
+        rdata.periodFinish = uint40(block.timestamp + REWARD_DURATION);
+    }
+
+    function _lastTimeRewardApplicable(uint256 _finishTime) internal view returns (uint256) {
+        if (_finishTime < block.timestamp) {
+            return _finishTime;
+        }
+        return block.timestamp;
+    }
+
+    function _distributeGold() internal {
+        /// @dev no op silent fail if nothing to distribute
+        ITempleGold(address(rewardToken)).mint();
+    }
+
+    function _voteWeight(address _account) private view returns (uint256) {
+        /// @dev Vote weights are always evaluated at the end of last week
+        /// Borrowed from st-yETH staking contract (https://etherscan.io/address/0x583019fF0f430721aDa9cfb4fac8F06cA104d0B4#code)
+        uint256 currentWeek = (block.timestamp / WEEK_LENGTH) - 1;
+        AccountWeightParams storage weight = _weights[_account];
+        uint256 week = uint256(weight.weekNumber);
+        if (week > currentWeek) {
+            weight = _prevWeights[_account];
+        }
+        uint256 t = weight.stakeTime;
+        uint256 updated = weight.updateTime;
+        if (week > 0) {
+            t += (block.timestamp / WEEK_LENGTH * WEEK_LENGTH) - updated;
+        }
+        return _balances[_account] * t / (t + halfTime);
+    }
+
+    function _updateAccountWeight(address _account, uint256 _prevBalance, uint256 _newBalance, bool _increment) private {
+        uint256 currentWeek = block.timestamp / WEEK_LENGTH;
+        uint256 week = 0;
+        uint256 t = 0;
+        uint256 updated = 0;
+        AccountWeightParams storage weight = _weights[_account];
+        (week, t, updated) = _unpackWeight(_account);
+        uint256 _lastShares = _prevBalance;
+        if (week > 0 && currentWeek > week) {
+            _prevWeights[_account] = weight;
+        }
+        if (_newBalance == 0) {
+            t = 0;
+            _lastShares = 0;
+        }
+        if (_lastShares > 0) {
+            t = t + block.timestamp - updated;
+            if (_increment) {
+                // amount has increased, calculate effective time that results in same weight
+                uint256 _halfTime = halfTime;
+                t = _prevBalance * t * _halfTime / (_newBalance * (t + _halfTime) - _prevBalance * t);
+            }
+        }
+        // update weight
+        weight.weekNumber = uint64(currentWeek);
+        weight.stakeTime = uint64(t);
+        weight.updateTime = uint64(block.timestamp);
+    }
+
+    function _unpackWeight(address _account) private view returns (uint256 week, uint256 t, uint256 updated) {
+        AccountWeightParams memory params = _weights[_account];
+        week = params.weekNumber;
+        t = params.stakeTime;
+        updated = params.updateTime;
+    }
+
+    modifier updateReward(address _account) {
```

Comment:
What's the effect of this not being called within a weekly period? Eg if no one stakes/unstakes/claims for 3 weeks are some rewards never claimable by users?

Pre-empting the audit, might be worth documenting assumptions on this if there is some behaviour like that.

Reply:
when rewards are distributed for next 7 days, the `rewardPerTokenStored` is updated. and this is useful in calculating stakers earned rewards. If this is not called, there are no rewards for the next 7 days.
`distrubuteRewards()` is expected to be called at  latest every 7 days, then such scenario is avoidable. it can also be called every day.
Stakers can always claim their rewards from past epochs.